torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -7513,3 +7513,39 @@ def test_ddp_bwd_with_retain_graph_static_single_fwd(self):
                 [param.grad for param in model_local.parameters()]
             )
             self.assertEqual(dist_grad_tensor, local_grad_tensor)
+
+        @unittest.skipIf(
+            BACKEND != "nccl" and BACKEND != "gloo",
+            "MPI backend does not support DDP communication hook on CUDA devices",
+        )
+        @skip_if_lt_x_gpu(2)
+        @skip_if_rocm
+        def test_hook_grad_idx_to_bucket_mapping(self):
+            # Test hook to ensure get_grad_index_to_variable_mapping works appropriately.
+            def _test_hook(hook_state, bucket: dist.GradBucket) -> torch.futures.Future:
+                # Run allreduce and wait for it to complete
+                allreduce_fut = default._allreduce_fut(None, bucket.get_tensor())
+                allreduce_fut.wait()
+                per_param_grad_tensors = bucket.get_per_parameter_tensors()
+                idx_to_model_params = bucket.get_grad_index_to_variable_mapping()
+                # Verify mapping is correct by ensuring param.grad is equal to
+                # the grad we get using the mapping.
+                for grad_tensor_idx, model_param in idx_to_model_params.items():
+                    self.assertEqual(model_param.grad, per_param_grad_tensors[grad_tensor_idx])
```

Comment:
If the above assumption is correct, how does this work? Since the test doesn't enable `grad_as_bucket_view`, I assume `param.grad` are not bucket views, and the allreduce result are not yet copied back to `param.grad`. I might miss sth, why do we expect `param.grad` and bucket views to be the same value?

Reply:
This is very strange because we're also not calling `zero_grad` anywhere so theoretically the grad field should accumulate. Looking into it...

Reply:
@mrshenli This appears to pass because the gradients are so close to 0 (on the order of 1e-8) that when I print them out they are indeed different, but `self.assertEqual` default tolerance is too generous marking them as equal.

If we scale up the gradients this test indeed fails. This makes sense because during backwards the grads have not yet been copied back to param.grad field. Although, the main use case here is to just verify we can get the appropriate gradient for the parameter, so we can probably just do shape check when `grad_as_bucket_view=False` and grad check when `grad_as_bucket_view=True`?

Reply:
Yep, that looks good to me.